### PR TITLE
Handle tun fd from native world

### PR DESCRIPTION
**Pull request by wongsyrone**

### CODE REVIEW

File: tun2socks/tun2socks.go [hunk 1]
```diff
@@ -1,74 +1,137 @@
...
 	"log"
 	"net"
+	"os"
+	"time"
 
 	"github.com/eycorsican/go-tun2socks/common/dns/cache"
 	"github.com/eycorsican/go-tun2socks/common/dns/fakedns"
 	"github.com/eycorsican/go-tun2socks/core"
 	"github.com/eycorsican/go-tun2socks/proxy/socks"
+	"github.com/songgao/water"
```
Comment (oasiscifr) [inline tun2socks/tun2socks.go]: 好像只用了water的数据结构，可以复制进来去掉这个依赖吗？

**Review by oasiscifr — APPROVED** [review]

File: tun2socks/tun2socks.go [hunk 2]
```diff
@@ -1,74 +1,137 @@
...
+	tunDev            *water.Interface
+)
 
 // Stop stop it
 func Stop() {
+	log.Printf("enter stop")
+	if stopSignalChannel != nil {
+		log.Printf("send stop sig")
+		close(stopSignalChannel)
+		time.Sleep(2 * time.Second)
```
Comment (oasiscifr) [inline tun2socks/tun2socks.go]: Why two seconds?

File: tun2socks/tun2socks.go [hunk 3]
```diff
@@ -1,74 +1,137 @@
...
+	Loop:
+		for {
+			select {
+			case <-c:
+				log.Printf("got DataPipe stop signal")
+				break Loop
+
+			default:
+				// tun -> lwip
+				log.Printf("start copy buffer")
```
Comment (oasiscifr) [inline tun2socks/tun2socks.go]: too many logs?

File: tun2socks/tun2socks.go (hunk 2 above)
Comment (wongsyrone) [reply tun2socks/tun2socks.go]: Make sure before Stop returned, the goroutine is terminated

File: tun2socks/tun2socks.go (hunk 3 above)
Comment (wongsyrone) [reply tun2socks/tun2socks.go]: Nope. This line appears only once.

**Comment by wongsyrone** [comment]
Refactored to release resources properly

**Review by oasiscifr — APPROVED** [review]

**Comment by oasiscifr** [comment]
lgtm, great work!

**Pull request merged**